### DISCUSSION TITLE
Benchmarks for GPU speedup?

### DISCUSSION THREAD

**Discussion by raguiar2** [idaholab/moose#29030] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Is there a benchmark on how much faster GPU-based simulation is for different solvers? all I see in the [official thread](https://github.com/idaholab/moose/discussions/21152) is that it will 'significantly speed up the simulation'.

**Comment by GiudGiud** [comment]
Hello

This is work in progress. We are currently able to assemble the system on the CPU and solve it on the GPU. Our next step is integrating both on the GPU.
What do you need the benchmark for?

Guillaume

**Reply by raguiar2** [reply]
I am choosing a tool for a set of Thermo-Mechanical Solves, and I wanted to test both MOOSE and MFEM on GPU. Just wanted to know if there was a baseline performance profile I could work off of. 

Also happy to get involved in the discussion of integrating both on the GPU, if there's a thread you can point me to.

**Reply by GiudGiud** [reply, marked as answer]
I think you'll find a lot more implemented CPU capability here than you will find with MFEM on the GPU. It depends of what you need in terms of physics & models. 

There is no public project afaik, GPU support is being developed internally.

**Reply by PengWei97** [reply]
So I would like to ask, will the GPU computing version based on MOOSE be made public? If yes, when will it be made public?

Because when I used phase field module, I found that the computing efficiency of GPU can reach 100~1000 times that of GPU.

wei

**Reply by GiudGiud** [reply]
There is a pull request open now for integrating MFEM into moose. This will be our GPU capability

**Reply by PengWei97** [reply]
Oh, that's great. Looking forward to having GPU acceleration in MOOSE. Could you provide the link of this pull request, so I can follow the development progress?

**Reply by GiudGiud** [reply]
https://github.com/idaholab/moose/pull/29633

**Reply by PengWei97** [reply]
Hi @GiudGiud,

I see that the branch you provided has been merged into idaholab:next, so does MOOSE have the GPU acceleration function now?
![image](https://github.com/user-attachments/assets/84f96fa3-60b2-472b-a2c7-99873c6830b3)
![image](https://github.com/user-attachments/assets/ae8c97ab-ea90-41ab-8322-8017932a0f48)

Wei

**Reply by GiudGiud** [reply]
hello

yes you can use GPUs with MOOSE through MFEM. This is a whole new capability, that does not rely on the existing objects, so you d have to use only the new objects to do that.
let me see if we have a tutorial

**Reply by PengWei97** [reply]
That's great. Thanks you for your reply. 

I hope that after using GPU acceleration, the computational efficiency of phase field simulation can be improved. Now I have two further questions:
1. Can the C++ class written based on CPU parallelism be directly migrated for GPU acceleration without re-modification?
2. Regarding the use of GPU computing, does MOOSE have any relevant tutorials? Please forgive me for only knowing that MOOSE can perform GPU computing, but I don't know how to operate it specifically.

Wei

**Reply by GiudGiud** [reply]
> Can the C++ class written based on CPU parallelism be directly migrated for GPU acceleration without re-modification?

no they need to be re-written using the MFEM model for kernels. 
Same for materials

> Regarding the use of GPU computing, does MOOSE have any relevant tutorials?

I ll let you know when I hear back

**Reply by PengWei97** [reply]
Ok, thank you very much. Looking forward to your further news.

**Reply by lindsayad** [reply]
We might want to get these migrated over to the moose framework site (@alexanderianblair @nmnobre): https://aurora-multiphysics.github.io/platypus/examples/index.html. This may also be helpful: https://mfem.org/pdf/workshop24/23_Blair_Platypus.pdf

**Reply by PengWei97** [reply]
Great, thanks a lot for sharing! I'll go through the materials carefully and look forward to your suggestions in future conversations.

**Reply by PengWei97** [reply]
Hi @lindsayad, I saw in the Platypus Workshop 2024 slides ([link](https://mfem.org/pdf/workshop24/23_Blair_Platypus.pdf)) that "Upcoming – MOOSE/MFEM Transfers" and "Upcoming – Nonlinear Problems". 

Has this been implemented? Any docs or timeline if not?

Thanks for any details!

Best,
Wei

**Reply by alexanderianblair** [reply]
An initial implementation for enabling users to specify nonlinear MFEM problems (effectively using parsed functors of variables) is in progress our side at UKAEA/STFC - we're hoping to demonstrate it at the upcoming MFEM Community Workshop in September. 

MOOSE/MFEM transfers don't have any major blocker to finish them off that I know of now other than bandwidth/prioritisation of other tasks (at least for CopyTransfers) I'd probably estimate EOY unless I can return to it sooner!
